w3d_translator/read_xml.py

```diff
@@ -50,51 +49,74 @@ def read_xml(file):
     return story
 
 
-def parse_camera(xml: ET.Element) -> dict:
-    return {
-        "far_clip": float(xml.attrib["far-clip"]),
-        "placement": parse_placement(xml.find("Placement")),
-    }
+def parse_string(string: str) -> Union[bool, int, float, tuple, str]:
+    # Check if string is a boolean
+    try:
+        test = string.lower()
+        if test == "true":
+            return True
+        elif test == "false":
+            return False
+    except AttributeError:
+        pass
 
+    # Check if string is an integer
+    try:
+        return int(string)
+    except ValueError:
+        pass
 
-def parse_placement(xml: ET.Element) -> dict:
-    relative_to = xml.find("RelativeTo").text
-    position = str_to_tuple(xml.find("Position").text)
-    axis = None
-    look_at = None
-
-    axis_xml = xml.find("Axis")
-    if axis_xml is not None:
-        axis = {
-            "rotation": str_to_tuple(axis_xml.attrib["rotation"]),
-            "angle": float(axis_xml.attrib["angle"]),
-        }
-
-    look_at_xml = xml.find("LookAt")
-    if look_at_xml is not None:
-        axis = {
-            "target": str_to_tuple(look_at_xml.attrib["target"]),
-            "up": str_to_tuple(look_at_xml.attrib["up"]),
-        }
+    # Check if string is a float
+    try:
+        return float(string)
+    except ValueError:
+        pass
 
-    return {
-        "relative_to": relative_to,
-        "position": position,
-        "Axis": axis,
-        "LookAt": look_at,
-    }
+    # Check if string is a tuple (of integers or floats)
+    test = string.replace("(", "").replace(")", "").split(",")
+    try:
+        return tuple([int(x) for x in test])
+    except ValueError:
+        pass
+    try:
+        return tuple([float(x) for x in test])
+    except ValueError:
+        pass
 
+    # Plain text
+    return string
 
-def parse_wand_navigation(xml: ET.Element) -> dict:
+
+def parse_attributes(xml: ET.Element) -> dict:
+    attributes = xml.attrib
+    for key, value in attributes.items():
+        attributes[key] = parse_string(value)
+    return attributes
```

Comment:
in addition to having the parse_attributes, you could also have `parse_xml` which recurses on each of the children and `parse_xml`'s and then also parses attributes on the xml as well.

It still feels to me like you're explicitly walking the tree by finding things instead of just walking the tree and doing a rough pass at parsing which will get you to mostly the same place as far as I can tell

Reply:
There are definitely some things I want to change about the structure on the first pass. For example a `<Sound>` looks like

```xml
              <Sound name="gertrubience" filename="./audio/gertrubience.mp3" autostart="false">
                     <Mode>
                            <Fixed/> <!-- or <Positional /> -->
                     </Mode>
                     <Repeat>
                            <NoRepeat/> <!-- or <Repeat /> -->
                     </Repeat>
                     <Settings freq="1.0" volume="0.4" pan="0.0"/>
              </Sound>
```

I'd like to present this as one dict where `fixed` is a bool, `repeat` is a bool, and freq/volume/pan are indivdual keys (as opposed to a "Settings" key with the freq/volume/pan as a child dict). The children in the `<Global>` root aren't that bad but mostly everywhere else the xml has a really unintuitive structure. I don't want to recurse totally blindly - I'm going to have to walk the tree to lint this kind of stuff no matter what.

Reply:
I think it's fine to escape the default parsing path when needed, but I'd start with a generic parser and then add escape hatches

Reply:
So a recursive function that checks for specific tags?

Reply:
you know the code best - I'd just try to keep things generic as much as possible

Reply:
I did add a recursive function call but I'll be somewhat selective when using it